erts/emulator/nifs/unix/unix_prim_file.c

```diff
@@ -1085,3 +1085,119 @@ posix_errno_t efile_altname(ErlNifEnv *env, const efile_path_t *path, ERL_NIF_TE
 
     return ENOTSUP;
 }
+
+posix_errno_t efile_copy_range_int(int fd_in, int fd_out, off64_t length) {
+    off64_t ret;
+
+#ifdef HAVE_COPY_FILE_RANGE
+    ret = syscall(SYS_copy_file_range, fd_in, NULL, fd_out, NULL, length, 0);
+
+    if (ret < 0) {
+        return errno;
+    }
+#else 
+    char buffer[4096];
+    off64_t bytes_copied = 0;
+    ssize_t bytes_read, bytes_written; 
+    
+    while (bytes_copied < length && (bytes_read = read(fd_in, buffer, sizeof(buffer))) > 0) {
+        ssize_t total_bytes_written = 0;
+
+        do {
+            bytes_written = write(fd_out, buffer + total_bytes_written, bytes_read - total_bytes_written);
+            if (bytes_written >= 0) {
+                total_bytes_written += bytes_written;
+                bytes_copied += bytes_written;
+            } else {
+                if (errno != EINTR) {
+                    return errno;
+                }
+            }
+        } while (total_bytes_written < bytes_read && bytes_copied < length);
```

Comment:
If you ask to read `MIN(length, sizeof(buffer)` bytes, you can skip `bytes_copied < length` here.

Reply:
good point, not thought about that, will do